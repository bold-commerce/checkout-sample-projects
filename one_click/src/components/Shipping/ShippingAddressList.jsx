import React, { useEffect, useMemo, useState } from 'react';
import NewShippingAddress from './NewShippingAddress';
import ShippingAddressItem from './ShippingAddressItem';

const ShippingAddressList = ({ addresses, onChange, selectedAddress, disabled }) => {
  const [prevAddress, setPrevAddress] = useState(null);

  const shippingAddressList = addresses.map((address) => (
    <ShippingAddressItem
      address={address}
      key={address.id}
      onChange={onChange}
      selectedAddress={selectedAddress}
      disabled={disabled}
    />
  ));

  useEffect(() => {
    if (selectedAddress) {
      setPrevAddress(selectedAddress);
    }
  }, [selectedAddress]);

  const lastSelectedAddress = useMemo(() => {
    if (prevAddress) {
      return addresses.find((address) => address.id === prevAddress);
    } else {
      return addresses[0];
    }
  }, [prevAddress])

  // TODO: On select of new address, remove old address from order
  return (
    <ul className="AddressList">
      {shippingAddressList}
      <NewShippingAddress
        selected={selectedAddress === null || !selectedAddress}
        onChange={onChange}
        disabled={disabled}
        defaultAddress={lastSelectedAddress}
      />
    </ul>
  )
};

export default ShippingAddressList;
