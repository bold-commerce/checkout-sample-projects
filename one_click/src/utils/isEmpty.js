const isEmpty = (obj) => {
  return JSON.stringify(obj) === JSON.stringify({});
};

export default isEmpty;