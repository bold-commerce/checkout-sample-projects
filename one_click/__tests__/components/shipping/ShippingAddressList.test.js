import React from 'react';
import ShippingAddressList from '../../../src/components/Shipping/ShippingAddressList';
import { render } from '@testing-library/react';
import { exampleShippingState as MOCKexampleShippingState } from '../../utils/shippingLinesHelper';
import {
    exampleAddress,
    exampleSavedAddresses,
    countries as MOCKcountries,
    caProvinces as MOCKprovinces,
    exampleAddress as MOCKexampleAddress
 } from '../../utils/addressHelpers';

jest.mock('@boldcommerce/checkout-react-components', () => ({
    ...jest.requireActual('@boldcommerce/checkout-react-components'),
    useShippingLines: () => ({
        data: MOCKexampleShippingState,
        loadindStatus: 'fulfilled',
        errors: [],
        updateShippingLines: (() => {})
    }),
    useCountryInfo: () => ({
        data: {
            countries: MOCKcountries,
            provinceLabel: "province",
            provinces: MOCKprovinces,
            showPostalCode: true,
            showProvince: true
        }
    }),
    useShippingAddress: () => ({
        data: MOCKexampleAddress,
        errors: [],
        submitShippingAdderess: (() => {})
    })
}));

describe('ShippingAddressList', () => {
    test('renders ShippingAddressList component with addresses', () => {
        const { asFragment } = render(
                <ShippingAddressList
                    addresses={exampleSavedAddresses}
                    onChange={null}
                    selectedAddress={exampleAddress}
                    disabled={null}
                />
        );
        expect(asFragment()).toMatchSnapshot();
    });

    test('renders ShippingAddressList component with onChange', () => {
        const { asFragment } = render(
                <ShippingAddressList
                    addresses={exampleSavedAddresses}
                    onChange={() => {}}
                    selectedAddress={exampleAddress}
                    disabled={null}
                />
        );
        expect(asFragment()).toMatchSnapshot();
    });

    test('renders ShippingAddressList component disabled', () => {
        const { asFragment } = render(
                <ShippingAddressList
                    addresses={exampleSavedAddresses}
                    onChange={null}
                    selectedAddress={exampleAddress}
                    disabled
                />
        );
        expect(asFragment()).toMatchSnapshot();
    });
});