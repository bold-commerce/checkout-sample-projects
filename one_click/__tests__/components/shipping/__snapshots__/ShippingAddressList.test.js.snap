// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShippingAddressList renders ShippingAddressList component disabled 1`] = `
<DocumentFragment>
  <ul
    class="AddressList"
  >
    <li>
      <div
        class="RadioItem"
      >
        <input
          disabled=""
          id="address_1"
          name="shipping_address"
          type="radio"
        />
        <label
          class="RadioItem__Content"
          for="address_1"
        >
          <p>
            John Doe
          </p>
          <p>
            , , Winnipeg, Manitoba, , Canada
          </p>
          <p />
        </label>
      </div>
    </li>
    <li>
      <div
        class="RadioItem"
      >
        <input
          disabled=""
          id="address_2"
          name="shipping_address"
          type="radio"
        />
        <label
          class="RadioItem__Content"
          for="address_2"
        >
          <p>
            John Doe
          </p>
          <p>
            , , Winnipeg, Manitoba, , Canada
          </p>
          <p />
        </label>
      </div>
    </li>
    <li>
      <div
        class="RadioItem"
      >
        <input
          disabled=""
          id="address_3"
          name="shipping_address"
          type="radio"
        />
        <label
          class="RadioItem__Content"
          for="address_3"
        >
          <p>
            John Doe
          </p>
          <p>
            , , Winnipeg, Manitoba, , Canada
          </p>
          <p />
        </label>
      </div>
    </li>
    <li>
      <div
        class="RadioItem"
      >
        <input
          disabled=""
          id="address_0"
          name="shipping_address"
          type="radio"
        />
        <label
          class="RadioItem__Content"
          for="address_0"
        >
          <p>
            Use new address
          </p>
        </label>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`ShippingAddressList renders ShippingAddressList component with addresses 1`] = `
<DocumentFragment>
  <ul
    class="AddressList"
  >
    <li>
      <div
        class="RadioItem"
      >
        <input
          id="address_1"
          name="shipping_address"
          type="radio"
        />
        <label
          class="RadioItem__Content"
          for="address_1"
        >
          <p>
            John Doe
          </p>
          <p>
            , , Winnipeg, Manitoba, , Canada
          </p>
          <p />
        </label>
      </div>
    </li>
    <li>
      <div
        class="RadioItem"
      >
        <input
          id="address_2"
          name="shipping_address"
          type="radio"
        />
        <label
          class="RadioItem__Content"
          for="address_2"
        >
          <p>
            John Doe
          </p>
          <p>
            , , Winnipeg, Manitoba, , Canada
          </p>
          <p />
        </label>
      </div>
    </li>
    <li>
      <div
        class="RadioItem"
      >
        <input
          id="address_3"
          name="shipping_address"
          type="radio"
        />
        <label
          class="RadioItem__Content"
          for="address_3"
        >
          <p>
            John Doe
          </p>
          <p>
            , , Winnipeg, Manitoba, , Canada
          </p>
          <p />
        </label>
      </div>
    </li>
    <li>
      <div
        class="RadioItem"
      >
        <input
          id="address_0"
          name="shipping_address"
          type="radio"
        />
        <label
          class="RadioItem__Content"
          for="address_0"
        >
          <p>
            Use new address
          </p>
        </label>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`ShippingAddressList renders ShippingAddressList component with onChange 1`] = `
<DocumentFragment>
  <ul
    class="AddressList"
  >
    <li>
      <div
        class="RadioItem"
      >
        <input
          id="address_1"
          name="shipping_address"
          type="radio"
        />
        <label
          class="RadioItem__Content"
          for="address_1"
        >
          <p>
            John Doe
          </p>
          <p>
            , , Winnipeg, Manitoba, , Canada
          </p>
          <p />
        </label>
      </div>
    </li>
    <li>
      <div
        class="RadioItem"
      >
        <input
          id="address_2"
          name="shipping_address"
          type="radio"
        />
        <label
          class="RadioItem__Content"
          for="address_2"
        >
          <p>
            John Doe
          </p>
          <p>
            , , Winnipeg, Manitoba, , Canada
          </p>
          <p />
        </label>
      </div>
    </li>
    <li>
      <div
        class="RadioItem"
      >
        <input
          id="address_3"
          name="shipping_address"
          type="radio"
        />
        <label
          class="RadioItem__Content"
          for="address_3"
        >
          <p>
            John Doe
          </p>
          <p>
            , , Winnipeg, Manitoba, , Canada
          </p>
          <p />
        </label>
      </div>
    </li>
    <li>
      <div
        class="RadioItem"
      >
        <input
          id="address_0"
          name="shipping_address"
          type="radio"
        />
        <label
          class="RadioItem__Content"
          for="address_0"
        >
          <p>
            Use new address
          </p>
        </label>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
